sky/serve/autoscalers.py

```diff
@@ -315,3 +342,139 @@ def _get_replica_ids_to_scale_down(num_limit: int) -> List[int]:
         if not scaling_options:
             logger.info('No scaling needed.')
         return scaling_options
+
+    @classmethod
+    def from_spec(
+            cls,
+            spec: 'service_spec.SkyServiceSpec') -> 'RequestRateAutoscaler':
+        # TODO(MaoZiming): use NAME to get the class.
+        if spec.use_fallback:
+            return FallbackRequestRateAutoscaler(spec)
+        else:
+            return RequestRateAutoscaler(spec)
+
+
+class FallbackRequestRateAutoscaler(RequestRateAutoscaler):
+    """FallbackRequestRateAutoscaler: Use spot with
+    on-demand as fallback to autoscale based on request rate.
+    """
+
+    def __init__(self, spec: 'service_spec.SkyServiceSpec') -> None:
+        super().__init__(spec)
+        self.base_ondemand_fallback_replicas: int = (
+            spec.base_ondemand_fallback_replicas
+            if spec.base_ondemand_fallback_replicas is not None else 0)
+        self.dynamic_ondemand_fallback: bool = (
+            spec.dynamic_ondemand_fallback
+            if spec.dynamic_ondemand_fallback is not None else False)
+
+    def update_version(self, version: int,
+                       spec: 'service_spec.SkyServiceSpec') -> None:
+        super().update_version(version, spec)
+        self.base_ondemand_fallback_replicas = (
+            spec.base_ondemand_fallback_replicas
+            if spec.base_ondemand_fallback_replicas is not None else 0)
+        self.dynamic_ondemand_fallback = (spec.dynamic_ondemand_fallback
+                                          if spec.dynamic_ondemand_fallback
+                                          is not None else False)
```

Comment:
(1) If the spec.base_ondemand_fallback_replicas and dynamic_ondemand_fallback are both set to None, it will create `RequestRateAutoscaler`. 
https://github.com/skypilot-org/skypilot/blob/serve-spot-no-placer/sky/serve/autoscalers.py#L347-L354

(2) When updating the service, it will replace `FallbackRequestRateAutoscaler` with `RequestRateAutoscaler`, so it won't be launching spot instances.
https://github.com/skypilot-org/skypilot/blob/serve-spot-no-placer/sky/serve/controller.py#L127-L131

Reply:
I see. In that case, should we assert for at least one of the variables should be specified in `FallbackRequestRateAutoscaler`?

Reply:
Agreed we should assert for at least one of the variables specified